### DISCUSSION TITLE
redesign before flakes stabilize: 1.0 #121

### DISCUSSION THREAD

**Discussion by nrdxp** [nrdxp/nixdev#36] [Ideas]
If there is anything about this repository that users don't like or feel could be simplified, we probably want to address those concerns before flakes become the default mechanism in NixOS. Before opening an issues, I'd like to start an informal discussion for any ideas or enhancements anyone would like to suggest.

**Comment by blaggacao** [comment]
Carving out my feedback on #206 here:

- [`extern` &rarr; make it top level api instead of a nested api object, that is, expose those in `flake.nix`](https://github.com/divnix/devos/pull/206#discussion_r600237678)
- [`suites/default.nix` &rarr; as an aggregate over profiles, those could live a) in `flake.nix` or b) `suites.nix` and only if they grow too complex (which they ever will?) users can stow them c) in `suites/default.nix`.](https://github.com/divnix/devos/pull/206#discussion_r602231945)


/* We should acknowledge that _flake.nix_ will be the first place people look to, so after being stripped down via [devos-as-a-library](https://github.com/divnix/devos/issues/214#issue-842024114), we need to make sure this file properly represents an index and self-evident semantic overview of what is going on in the repo.

**Comment by blaggacao** [comment]
On a related note, I think a compact well-designed library is mentally cheaper than multiple "libraries" based on each concept (host, modules, suites) in a sub-flake.

I think so, because devos &mdash; all in all &mdash; is a coherent set of concepts.

**Comment by nrdxp** [comment]
The only reason I support subflakes isn't to try and make sublibraries. It is because there is absolutely no computation allowed on existing inputs. Meaning with just a single flake, you are forced to declare all your inputs in a single place. As the curent flake.nix shows, this gets pretty unmanagable pretty quickly. 

If we could get nix to resolve this upstream with something like what I've mentioned in [NixOS/nix#3978](https://github.com/NixOS/nix/issues/3978#issuecomment-795927581), we could ensure that the subflakes are kept as private information, by disallowing `?dir` directives from flakes explicitly imported with `self:path-to-subflake`.

I'm actually thinking of implementing this myself as my first addition to upstream nix.

**Reply by blaggacao** [reply]
I guess, I've to better understand the use case, still. Some inaccurate rough image might have built up in my head.

Is the main use case `./pkgs` then to declare sources?

https://github.com/NixOS/nix/issues/3978#issuecomment-795927581
